Task 2 - Clean repo setup with Webpack

davidverweij opened the pull request:
This PR (branched of PR #1) empties the complete repo, and set's up a new (_Typescript_) React project using `webpack` and `babel`. Using `webpack-merge`, three `webpack.config` files are used to create a `dev` environment (`npm start`) and a production build (`npm run build`).

@Dmitry-White, let me know if there are any improvements to be made!

davidverweij commented on webpack.prod.js:
```diff
@@ -0,0 +1,6 @@
+const common = require("./webpack.common.config");
+
+module.exports = merge(common, {
+  mode: "production",
```
That a very nice point, I've implemented it and the app (even though very empty at the moment!) can now be seen on [davidverweij.com/reactjs-practice-app/](https://www.davidverweij.com/reactjs-practice-app/)

davidverweij commented on package.json:
```diff
@@ -1,38 +1,42 @@
-    "test": "react-scripts test",
-    "eject": "react-scripts eject"
+    "start": "webpack serve --config webpack.dev.js",
+    "build": "webpack --config webpack.prod.js"
```
Got it - similar to your advice at the NodeJS MP (I have to remind myself adding these steps when setting up a project...)

davidverweij commented on src/components/App.tsx:
```diff
@@ -0,0 +1,15 @@
+import React from "react";
```
Done, I have implemented the plugin as suggested!

davidverweij commented:
Thanks @Dmitry-White! Very useful feedback and I completely agree - I will have to remind myself more about the best practices of setting up a project, such as including linting and alike.
I've adjusted the code base as suggested, let me know what you think.

Dmitry-White commented on package.json:
```diff
@@ -1,38 +1,62 @@
+    "eslint:fix": "eslint ./src --fix",
+    "pretty": "prettier --check .",
+    "pretty:fix": "prettier --write .",
+    "lint": "npm run eslint && npm run pretty",
```
Let's go a step further this time and add a pre-push hook with [Husky](https://typicode.github.io/husky/#/)

Dmitry-White reviewed (changes requested):
Also, optionally, think about adding Docker as well

davidverweij commented:
Cheers @Dmitry-White, I've added a pre-commit hook with `Husky` and set up for usage with Docker. I encountered some trade-offs with the Docker setup regarding the node_modules folder and whether to mount these to the container or not. _See the newly added changes to the README.md file with my approach_. To me, it seems illogical to depend on node_modules in the local folder, as this would render the 'custom' Docker image somewhat ineffective, but it seems a contested trade-off online. Any thoughts? Perhaps we can have a quick chat about this and your experiences with this? Cheers!

Dmitry-White reviewed (changes requested):
Creating a separate volume just to map host-specific node_modules is not a good idea.
A more easy solution would be to copy all (or selected) files to the container, install modules and leave only a hot-reload stuff as a volume (which is src folder, cause everything outside of src is affecting the build and therefore requires rebuild even when not using Docker).
<img width="1046" alt="Screenshot 2022-04-29 at 14 22 08" src="https://user-images.githubusercontent.com/25229823/165953374-f19b0949-2794-434c-9f04-c6765786166e.png">
<img width="900" alt="Screenshot 2022-04-29 at 14 22 12" src="https://user-images.githubusercontent.com/25229823/165953394-4878c077-d26e-4a2a-90d0-b0bd8ccc6594.png">

davidverweij commented:
Ah cheers, good solution - thanks for that! I've changed it, and also added docker commands as `npm scripts` to streamline the setup, build and push processes. It now takes into account the package/project version so that each branch/task/PR can pull the respective docker images based on the state of the MP project.

Dmitry-White reviewed (approved):
Looks awesome now!
